design-proposals/isolate-emulator-thread-with-smt-fix/isolate-emulator-thread-with-smt-fix.md

```diff
@@ -0,0 +1,217 @@
+3. VM is configured with `isolateEmulatorThread: true`
+
+then the VMI mutating-webhook will
+1. Copy the annotation from the kubevirt CR to the VMI annotations. This is done in order for it to be used on the virt-launcher, when setting the housekeeping cgroup. Doing so avoids adding API on the virt-handler to virt-launcher grpc protocol.
+
+## Virt-controller:
+When the following conditions are met:
+1. kubevirt CR has the annotation.
+2. kubevirt CR enabled the feature gate.
+3. VM is configured with `isolateEmulatorThread: true`
+
+then virt-controller will:
+1. Set the virt-launcher pod with an even amount of CPUs, meaning:
+
+| User requested CPUs               |Even (2,4,6,8,...)| odd (1,3,4,5,7,9,...) | Notes                                                                       |
+|-----------------------------------|------------------|-----------------------|-----------------------------------------------------------------------------|
+| Extra emulatorThread CPUs for pod |        +2        | +1 (as usual)         | * In both scenarios the guest will get the same amount of CPUs it requested |
+
+## Virt-launcher:
+When the following conditions are met:
+1. VMI has the annotation.
+2. VMI is configured with `isolateEmulatorThread: true`
+
+then virt-launcher pod will then:
+1. Assign the emulator Threads (one or two, depending on the User requested CPUs as shown in the table above) to the domain's emulatorThreadPin
+
+## Virt-handler:
+After virt-launcher set the emulatorThreadPin, the virt-handler will then:
+1. Set the emulator Threads from the emulatorThreadPin to the housekeeping cgroup.
+
+Pros:
+* Cleaner user experience for the VM owner. No prior knowledge of CPUs added behind the scene is needed.
+* VM configuration is less complex.
+* As this is an alpha annotation, it can be removed/ignored when and if solutions inside/outside Kubevirt such as CPU Manager handles the situation differently.
+
+Cons:
+* Setting an annotation in kubevirt CR is unconventional. 
+* A cluster-wide configuration such as this may be CPU-wasteful for heterogeneous clusters, where node-level granularity is needed.
+
+# Fix Alternatives
+## Alternative #1 Adding `EmulatorThreadEvenParity` annotation per VMI
+Another way to do it is by adding the annotation on the VMI's annotations:
+```
+apiVersion: kubevirt.io/v1
+kind: VirtualMachine
+spec:
+  running: true
+  template:
+    metadata:
+      annotations:
+        kubevirt.io/EmulatorThreadEvenParity:
+    spec:
+      domain:
+        cpu:
+          dedicatedCpuPlacement: true
+          isolateEmulatorThread: true
+          sockets: 1
+          cores: 4
+          threads: 2
+```
+Setting this will set the same behavior as on the suggested solution on virt-launcher and virt-handler.
+As opposed to the suggested solution where the user needs to add this annotation once on the kubevirt CR, it is possible to make this behavior VMI specific. This removes the assumption that the Nodes are homogeneous.
```

Comment:
```suggestion
As opposed to the suggested solution where the cluster admin needs to add this annotation once on the kubevirt CR, it is possible to make this behavior VMI specific. This would make it possible for owners of VM with even-numbered CPU, to run them on non-homogeneous clusters without wasting one CPU per VM.
```

There is no *requirement* to have a homogeneous cluster in either alternative. The only thing that different is who we are optimizing for.

Reply:
ACK. 
DONE.